lib/oppen/token_printer.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+# Oppen
+module Oppen
+  # TokenPrinter
+  class TokenPrinter
+    # PrintStackEntry
+    class PrintStackEntry
+      attr_reader :offset
+      attr_reader :break_type
+
+      def initialize(offset, break_type)
+        @offset = offset
+        @break_type = break_type
+      end
+    end
+
+    attr_reader :space
+
+    def initialize(line_width)
+      @print_stack = []
+      @space = @margin = line_width
+      @output = StringIO.new
+    end
+
+    def output
+      @output.string
+    end
+
+    def print_(token, token_length)
+      case token
+      when Token::Begin
+        print_handle_begin token, token_length
+      when Token::End
+        print_handle_end
+      when Token::Break
+        print_handle_break token, token_length
+      when Token::String
+        print_handle_string token, token_length
+      else
+        raise 'TokenPrinter print: Invalid Token'
+      end
+    end
+
+    def print_handle_begin(token, token_length)
+      if token_length > @space
+        type =
+          if token.break_type == Token::BreakType::CONSISTENT
+            Token::BreakType::CONSISTENT
+          else
+            Token::BreakType::INCONSISTENT
+          end
+        push PrintStackEntry.new @space - token.offset, type
+      else
+        push PrintStackEntry.new 0, Token::BreakType::FITS
+      end
+    end
+
+    def print_handle_end
+      pop
+    end
+
+    def print_handle_break(token, token_length)
+      block = top
+      case block.break_type
+      when Token::BreakType::FITS
+        @space -= token.blank_space
+        indent token.blank_space
+      when Token::BreakType::CONSISTENT
+        @space = block.offset - token.offset
+        new_line @margin - @space
+      when Token::BreakType::INCONSISTENT
+        if token_length > @space
+          @space = block.offset - token.offset
+          new_line @margin - @space
+        else
+          @space -= token.blank_space
+          indent token.blank_space
+        end
+      end
+    end
+
+    def print_handle_string(token, token_length)
+      if token_length > @space
+        raise 'TokenPrinter print_handle_string: Line too long'
+      end
+
+      @space -= token_length
+      puts_ token.value
+    end
+
+    def push(print_stack_entry)
+      @print_stack.append(print_stack_entry)
+    end
+
+    def pop
+      @print_stack.pop
+    end
+
+    def top
+      @print_stack.last
+    end
+
+    def new_line(amount)
+      puts_ '\n'
+      indent amount
+    end
+
+    def puts_(string)
+      @output.write(string)
+    end
+
+    def indent(amount)
+      puts_ ' ' * amount
+    end
+  end
+end
```

Comment:
You really want this class to act as an IO.
To achieve this, you must extend Forwardable, so you need to also understand extend vs include vs prepend.

1. https://dev.to/abbiecoghlan/ruby-modules-include-vs-extend-vs-prepend-4gmc
2. https://docs.ruby-lang.org/en/3.3/Forwardable.html

Now you can forward the IO methods to the underlying buffer, and override what you want.

It may be better to do it in a different iteration. I will make an issue here for it. But read them anyway.